File: src/main/java/com/milaboratory/mixcr/cli/ABaseCommand.java
package com.milaboratory.mixcr.cli;

import picocli.CommandLine;
import picocli.CommandLine.Option;
import picocli.CommandLine.Spec;

import java.util.stream.Collectors;

/**
 *
 */
public class ABaseCommand {
    @Spec
    public CommandLine.Model.CommandSpec spec; // injected by picocli

    @Option(names = {"-h", "--help"},
            hidden = true)
    public void requestHelp(boolean b) {
        throwValidationException("ERROR: -h / --help is not supported: use `mixcr help [command]` for command usage.");
    }

    /** Throws validation exception */
    public void throwValidationException(String message, boolean printHelp) {
        throw new ValidationException(spec.commandLine(), message, printHelp);
    }

    /** Throws validation exception */
    public void throwValidationException(String message) {
        throwValidationException(message, true);
    }

    /** Throws execution exception */
    public void throwExecutionException(String message) {
        throw new CommandLine.ExecutionException(spec.commandLine(), message);
    }

    public String getCommandLineArguments() {
        return spec.commandLine().getParseResult().originalArgs().stream().collect(Collectors.joining(" "));
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/ABaseCommand.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli

import picocli.CommandLine
import picocli.CommandLine.Option
import picocli.CommandLine.Spec

open class ABaseCommand(protected val appName: String) {
    // injected by picocli
    @Spec
    lateinit var spec: CommandLine.Model.CommandSpec

    @Suppress("unused", "UNUSED_PARAMETER")
    @Option(names = ["-h", "--help"], usageHelp = true)
    fun requestHelp(b: Boolean) {
    }

    val commandLineArguments: String
        get() = spec.commandLine().parseResult.originalArgs().joinToString(" ")
}
